### ISSUE TITLE
spurious NMS in region layer which is not compatible with darknet

### ISSUE THREAD

**Issue by YashasSamaga** [opencv/opencv#17415]
##### System information (version)
- OpenCV => 4.3.0

##### Detailed description

**History of the region layer:**

- YOLOv2 added in #9705. YOLOv2 detected objects at one scale; hence, there is just one region layer. The region layer would output unfiltered detections. The obvious step after region layer was to perform NMS. Hence, NMS was presumably integrated into the region layer.
- YOLOv3 support was added in #11322. The improved YOLOv3 was detecting objects at three different scales. Each scale had a separate yolo layer in darknet and NMS was performed on all the detections at the end. The #11322 creates a region layer for each yolo layer and *hence performs NMS in each region layer*.

**Is the nms spurious?**

The NMS was added in the YOLOv2 PR where there was just one region layer. The next step after the region computation would be NMS (which is probably why NMS was integrated into region). YOLOv3 has three region layers. Is it right to use the region layer as it is without disabling NMS?

**Performance penalty for the CUDA backend:**
The NMS is performed on CPU in the CUDA backend. This has a huge performance toll. The switch from GPU to CPU during inference introduces additional synchronization overhead, interference from OS scheduling decisions and prevents the latency hiding ability of CUDA streams.

The toll is so significant that FPS of YOLOv4 improves by almost 40% on RTX 2070 Super when NMS is disabled.

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [ ] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by dkurt** [comment]
According to experiment from https://github.com/opencv/opencv/pull/17332 NMS over three Region layers in YOLOv3 is required.

In https://github.com/opencv/opencv/pulls we turns nms_threshold OFF inside the layer to remove NMS duplication but applies separately over all the detections.

**Comment by YashasSamaga** [comment]
> According to experiment from #17332 NMS over three Region layers in YOLOv3 is required.

This is very confusing. From what I inferred from darknet source is that the detections of all the region layers are put together in one big array and NMS is performed on the whole lot in one go. I will have to check again.

**Comment by dkurt** [comment]
@YashasSamaga, Let me rephrase a bit. We just did an experiment to check if global NMS is required or it's enough to have NMS at Region layers. We can confirm that even NMS at Region is enabled global one is also required. That means we can disable it for Region layers from YOLOv3/v4 during import (i.e. detect if `mask` field is present). For now https://github.com/opencv/opencv/pull/17371 does it automatically for all the YOLOs but in case of `DetectionModel` interface.

**Comment by YashasSamaga** [comment]
> That means we can disable it for Region layers from YOLOv3/v4 during import (i.e. detect if mask field is present).

Is there any chance that this will be done before the next release?

**Comment by dkurt** [comment]
@YashasSamaga, sorry. I just thought that you closed issue as resolved.

**Comment by l-bat** [comment]
PR #17592 with a fix has been merged to 3.4 branch
